src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/AbstractReferenceHandler.cs

```diff
@@ -68,32 +71,73 @@ internal async Task<List<ProjectSystemReferenceInfo>> GetReferencesAsync(Configu
 
             var item = projectItems
-                .FirstOrDefault(c => c.EvaluatedInclude == referenceUpdate.ReferenceInfo.ItemSpecification);
+                .FirstOrDefault(c => c.EvaluatedInclude == itemSpecification);
+            return item;
+        }
+
+        internal IProjectSystemUpdateReferenceOperation CreateSetAttributeCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new SetAttributeCommand(this, selectedConfiguredProject, referenceUpdate.ReferenceInfo.ItemSpecification);
+        }
+
+        internal IProjectSystemUpdateReferenceOperation CreateUnSetAttributeCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new UnSetAttributeCommand(this, selectedConfiguredProject, referenceUpdate.ReferenceInfo.ItemSpecification);
+        }
+
+        internal IProjectSystemUpdateReferenceOperation? CreateRemoveReferenceCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new RemoveReferenceCommand(this, selectedConfiguredProject, referenceUpdate);
+        }
+
+        public async Task<Dictionary<string, string>> GetAttributesAsync(ConfiguredProject selectedConfiguredProject, string itemSpecification)
+        {
+            Dictionary<string, string> propertyValues = new ();
 
-            if (item != null)
+            IProjectItem items = await GetProjectItems(selectedConfiguredProject, itemSpecification);
+
+            var propertyNames = await items.Metadata.GetPropertyNamesAsync();
+
+            foreach (var property in propertyNames)
             {
-                string newValue = PropertySerializer.SimpleTypes.ToString(referenceUpdate.Action == ProjectSystemUpdateAction.SetTreatAsUsed);
+                var value = await items.Metadata.GetEvaluatedPropertyValueAsync(property);
+                propertyValues.Add(string.Copy(property), string.Copy(value));
+            }
 
-                await item.Metadata.SetPropertyValueAsync(ProjectReference.TreatAsUsedProperty, newValue, null);
+            return propertyValues;
+        }
 
-                wasUpdated = true;
+        public async Task SetAttributes(ConfiguredProject selectedConfiguredProject, string itemSpecification, Dictionary<string, string>? projectPropertiesValues)
+        {
+            if (projectPropertiesValues is null)
+            {
+                return;
             }
 
-            return wasUpdated;
+            IProjectItem items = await GetProjectItems(selectedConfiguredProject, itemSpecification);
+
+            if (items != null)
+            {
+                foreach (var property in projectPropertiesValues)
+                {
+                    await items.Metadata.SetPropertyValueAsync(property.Key, property.Value, null);
+                }
```

Comment:
It's possible to destructure here. I like the fact that it gives control over the name (rather than `Key`/`Value`).

```suggestion
                foreach ((string propertyName, string propertyValue) in projectPropertiesValues)
                {
                    await items.Metadata.SetPropertyValueAsync(propertyName, propertyValue, null);
                }
```